File: app_backend/src/routes/execute.js
import { Router } from "express";
import { requireAuth } from "../middleware/auth.js";
import { getSequelize } from "../config/database.js";

const router = Router();

/**
 * @openapi
 * /api/execute:
 *   post:
 *     tags: [Execute]
 *     security: [{ bearerAuth: [] }]
 *     summary: Execute an action with plan-gated behavior
 *     description: >
 *       Demonstrates conditional behavior based on the authenticated user's plan.
 *       Returns a different message and payload fields for normal, premium, and ultra plans.
 *     requestBody:
 *       required: false
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             additionalProperties: true
 *     responses:
 *       200:
 *         description: Execution result varies by plan
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 success: { type: boolean }
 *                 plan: { type: string, enum: [normal, premium, ultra] }
 *                 message: { type: string }
 *       401:
 *         description: Unauthorized
 */
router.post("/", requireAuth, async (req, res, next) => {
  try {
    const sequelize = getSequelize();
    const User = sequelize.models.User;
    const user = await User.findByPk(req.auth.uid);
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    const plan = user.plan;
    let payload = { success: true, plan, message: "" };

    // Conditional behavior by plan
    if (plan === "normal") {
      payload.message = "Basic execution complete. Upgrade for more features.";
      payload.limit = "low";
    } else if (plan === "premium") {
      payload.message = "Premium execution complete with enhanced processing.";
      payload.limit = "medium";
      payload.speed = "fast";
    } else if (plan === "ultra") {
      payload.message = "Ultra execution complete with all features unlocked.";
      payload.limit = "maximum";
      payload.speed = "fastest";
      payload.priority = true;
    } else {
      payload.message = "Unknown plan";
    }

    res.json(payload);
  } catch (err) {
    next(err);
  }
});

export default router;

File: app_backend/fastapi_app/routers/execute.py
from typing import Annotated, Any, Dict, Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from ..auth import get_current_user
from ..models import Plan, User

router = APIRouter(prefix="/execute", tags=["Execute"])


class ExecuteRequest(BaseModel):
    """Optional arbitrary payload for execution."""
    # Accept any additional fields
    payload: Optional[Dict[str, Any]] = Field(default=None, description="Arbitrary payload for execution")


class ExecuteResponse(BaseModel):
    """Response varies based on user's plan."""
    success: bool = True
    plan: Plan
    message: str
    limit: Optional[str] = None
    speed: Optional[str] = None
    priority: Optional[bool] = None


# PUBLIC_INTERFACE
@router.post(
    "",
    summary="Execute an action with plan-gated behavior",
    description=(
        "Demonstrates conditional behavior based on the authenticated user's plan. "
        "Returns a different message and payload fields for normal, premium, and ultra plans."
    ),
    response_model=ExecuteResponse,
    responses={401: {"description": "Unauthorized"}},
)
def execute(
    _req: ExecuteRequest | None = None,
    current_user: Annotated[User, Depends(get_current_user)] = None,  # type: ignore[assignment]
) -> ExecuteResponse:
    """Perform a dummy action; response content varies depending on the user's plan."""
    plan = current_user.plan  # type: ignore[union-attr]
    base = ExecuteResponse(success=True, plan=plan, message="")

    if plan == Plan.normal:
        base.message = "Basic execution complete. Upgrade for more features."
        base.limit = "low"
    elif plan == Plan.premium:
        base.message = "Premium execution complete with enhanced processing."
        base.limit = "medium"
        base.speed = "fast"
    elif plan == Plan.ultra:
        base.message = "Ultra execution complete with all features unlocked."
        base.limit = "maximum"
        base.speed = "fastest"
        base.priority = True
    else:
        base.message = "Unknown plan"

    return base
